### PR TITLE
DPOC-225 Category page

### PR DESCRIPTION
**Pull request by maksym-bielyshev**

### CODE REVIEW

File: dp189/components.py [hunk 1]
```diff
@@ -106,3 +110,392 @@ def click_transactions(self):
     def click_newsletter(self):
         self._right_menu.find_element(*RightMenuLocators.NEWSLETTER).click()
 
+
+class LeftCategoryMenuComponent:
```
Comment (tkruh) [inline dp189/components.py]: Proposal: Create single method (for example show_items_for_category (category)) for all categories in left menu which would accept text parameter "category name" and return the list of product widgets. Letter in the tests we would compere this list with known list of products in the category.

File: dp189/components.py [hunk 2]
```diff
@@ -106,3 +110,392 @@ def click_transactions(self):
...
+        self.compare_button = driver.find_element(By.XPATH, f"{self.product_title}/../../../div[2]/button[3]")
+
+    def click_add_to_shopping_cart_button(self) -> None:
+        """Click on the "Add to shopping cart" button.
+
+        :return: None
+        """
+        self.shopping_cart_button.click()
+
+    def click_add_to_wish_list_button(self) -> None:
```
Comment (tkruh) [inline dp189/components.py]: Consider the possibility to return of allert web element "success" in the top.

File: dp189/pages/category_page.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        """
+        super().__init__(driver)
+        self.category_title = driver.find_element_by_id("content")
+        self.left_category_menu = LeftCategoryMenuComponent(driver)
+        self.list_view = ListViewComponent(driver)
+        self.grid_view = GridViewComponent(driver)
+        self.product_compare_link = ProductCompareLinkComponent(driver)
+        self.sort_by = SortByDropdownComponent(driver)
+        self.show_number_products = ShowNumberProductsDropdownComponent(driver)
+        self.product = CategoryProductWidgetComponent(driver, None)
```
Comment (tkruh) [inline dp189/pages/category_page.py]: This should probably be a list of the product widget instances

File: dp189/components.py [hunk 4]
```diff
@@ -106,3 +110,392 @@ def click_transactions(self):
...
+
+    def click_mp3_players(self) -> None:
+        """Click on the 'MP3 Players' link.
+
+        :return: None
+        """
+        self._driver.find_element(*LocatorsLeftCategoryMenu.MP3_PLAYERS).click()
+
+
+class CategoryProductWidgetComponent:
```
Comment (tkruh) [inline dp189/components.py]: This widget object should be probably initialized with data read from web page?

File: dp189/components.py [hunk 5]
```diff
@@ -106,3 +110,392 @@ def click_transactions(self):
...
+
+    def click_add_to_compare_button(self) -> None:
+        """Click on the "Compare this Product" button.
+
+        :return: None
+        """
+        self.compare_button.click()
+
+
+class SortByDropdownComponent:
```
Comment (tkruh) [inline dp189/components.py]: Same as for LeftCategoryMenuComponent

File: dp189/components.py [hunk 6]
```diff
@@ -106,3 +110,392 @@ def click_transactions(self):
...
+    def click_model_z_a(self) -> None:
+        """Click on the sorting drop-down and on the 'Model (Z - A)' option.
+
+        :return: None
+        """
+        self.click_drop_down()
+        self._model_z_a.click()
+
+
+class ShowNumberProductsDropdownComponent:
```
Comment (tkruh) [inline dp189/components.py]: Same as for LeftCategoryMenuComponent

File: dp189/components.py [hunk 7]
```diff
@@ -106,3 +110,392 @@ def click_transactions(self):
...
+    def click_hundred(self) -> None:
+        """Click on the sorting drop-down and on the '100' option.
+
+        :return: None
+        """
+        self.click_drop_down()
+        self._hundred.click()
+
+
+class ProductCompareLinkComponent:
```
Comment (tkruh) [inline dp189/components.py]: Do we really need this class?

File: dp189/components.py [hunk 8]
```diff
@@ -106,3 +110,392 @@ def click_transactions(self):
...
+
+    def click_product_compare_link(self) -> None:
+        """Click on the 'Product Compare' link.
+
+        :return: None
+        """
+        self._product_compare_link.click()
+
+
+class ListViewComponent:
```
Comment (tkruh) [inline dp189/components.py]: In my opinion these grid and list elements are not finished yet. You can remove them for now and finish later.

File: dp189/pages/category_page.py (hunk 3 above)
Comment (maksym-bielyshev) [reply dp189/pages/category_page.py]: This component is for working with only one widget: pressing the buttons of a certain product.

File: dp189/components.py (hunk 7 above)
Comment (maksym-bielyshev) [reply dp189/components.py]: I think so because this element is also found on the other page ("search" page).

File: dp189/components.py (hunk 8 above)
Comment (maksym-bielyshev) [reply dp189/components.py]: What else needs to be done for these elements?

File: dp189/pages/category_page.py [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+class CategoryPage(BasePage):
+    """Category page class."""
+
+    def __init__(self, driver: Remote) -> None:
+        """All objects on product page.
+
+        :param driver: Remote
+        """
+        super().__init__(driver)
+        self.category_title = driver.find_element_by_id("content")
```
Comment (maksimok93) [inline dp189/pages/category_page.py]: I think that static locators must be in Locators.py module.

And, according to team convention, we should find elements in this way:
self.category_title = driver.find_element(By.ID, "content")

Comment (dianakardashova) [reply dp189/pages/category_page.py]: Maybe it must be like this:
self.category_title = driver.find_element(*LocatorsCategoryPage.CATEGORY_TITLE)

in locators.py:
class LocatorsCategoryPage:
    CATEGORY_TITLE = (By.ID, "content")

File: dp189/components.py [hunk 10]
```diff
@@ -116,6 +158,143 @@ def click_newsletter(self):
...
+
+    def __init__(self, driver: Remote) -> None:
+        """Initialise the link.
+
+        :param driver: Remote driver.
+        """
+        self._driver = driver
+        self.product_compare_link = driver.find_element(*LocatorProductCompareLink.PRODUCT_COMPARE)
+
+    def click_product_compare_link(self) -> None:
```
Comment (tkruh) [inline dp189/components.py]: This method should return ProductComperePage

File: dp189/components.py [hunk 11]
```diff
@@ -116,6 +158,143 @@ def click_newsletter(self):
...
+class ProductCompareLinkComponent:
+    """A link to compare the products added to the comparison."""
+
+    def __init__(self, driver: Remote) -> None:
+        """Initialise the link.
+
+        :param driver: Remote driver.
+        """
+        self._driver = driver
+        self.product_compare_link = driver.find_element(*LocatorProductCompareLink.PRODUCT_COMPARE)
```
Comment (tkruh) [inline dp189/components.py]: This line should be in method click_product_compare_link

**Review by maksimok93 — APPROVED** [review]

**Review by Danya1998 — APPROVED** [review]

**Review by dianakardashova — APPROVED** [review]

**Pull request merged**